File: LeetCode/234/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode *reverse(ListNode *node) {
    ListNode *curr = node;
    ListNode *prev = NULL;
    while (curr) {
      ListNode *nxt = curr->next;
      curr->next = prev;
      prev = curr;
      curr = nxt;
    }
    return prev;
  }
  bool isPalindrome(ListNode *head) {
    ListNode *ptr1 = head;
    ListNode *ptr2 = head;

    while (ptr2 && ptr2->next) {
      ptr1 = ptr1->next;
      ptr2 = ptr2->next->next;
    }

    ptr1 = reverse(ptr1);
    ptr2 = head;

    while (ptr1) {
      if (ptr1->val != ptr2->val) {
        return false;
      }
      ptr1 = ptr1->next;
      ptr2 = ptr2->next;
    }

    return true;
  }
};

File: LeetCode/234/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        def reverse(node):
            prev = None
            while node:
                nxt = node.next
                node.next = prev
                prev = node
                node = nxt
            return prev

        slow = fast = temp = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        slow = reverse(slow)

        while slow:
            if temp.val != slow.val:
                return False
            else:
                temp = temp.next
                slow = slow.next

        return True
